### DISCUSSION TITLE
Is there a way to execute raw sql from the javascript library?

### DISCUSSION THREAD

**Discussion by sanketch** [supabase/supabase#444] [Questions]
Wasn't able to find any documentation on this, but I am trying to see if the library supports the ability to run a custom SQL query that I can send over to execute? I am looking for Group by functionality to aggregate and realized that the library doesn't have a function for that. 

Or the only route for this is through stored procedures?

**Comment by inian** [comment]
Hi @sanketch, yes stored procedures are the way to go here. Postgrest [doesn't let you execute raw SQL](https://github.com/PostgREST/postgrest/issues/149#issuecomment-74619277) queries since they aren't RESTful.

**Reply by steve-chavez** [reply]
To further clarify, here's a link to the postgrest docs: https://postgrest.org/en/v7.0.0/api.html#custom-queries

Raw SQL is not supported through postgrest because it opens up a way for a malicious user to DoS your database. SQL is turing complete after all(recursive CTEs).

**Comment by steve-chavez** [comment]
Besides stored procedures, you can create a view for aggregates. Example:

```sql
create view public.projects_count_grouped_by as
select
  client_id,
  count(id) as number_of_projects
from projects
group by client_id;
```

Then you can use the view the same as any other table through `supabase-js`:

```sql
let { data, error } = await supabase
  .from('projects_count_grouped_by')
  .select('*')
```

**Comment by eifo** [comment]
If only we had an equivalent to prisma $queryRaw, we cannot pass parameters to views and creating functions and procedures is time consuming and harder to track and edit, not something I'm looking forward to do for every single query, leaves no option other than learn the syntax.

**Reply by kiwicopple** [reply]
Thanks for the input @eifo - I didn't know about Prisma's way of doing it. We'll dig into their [implementation](https://github.com/prisma/prisma/issues/3828) to see how they are handling security concerns on the client for raw data. 

Just a note - we will probably bring this out at some point, but views/procedures are still encouraged for performance.

**Reply by eifo** [reply]
It would be helpful in the meantime to have the functions procedures displayed and editable under "stored procedures" until then a secondary postgres client will be needed to view and update.

**Reply by eifo** [reply]
Hello @kiwicopple just revisiting this over here, is this something in the works to be able to do raw sql? The alternative of rpc and views are fine, but to be able to write raw sql directly would be so much faster.
